### ISSUE TITLE
Add a way to compile and run tests in examples

### ISSUE THREAD

**Issue by Bobo1239** [rust-lang/cargo#2631]
Creating a minimal example and running `cargo test --example test` yields no output (expect the 'Compiling ...' message on the first run).
The example gets compiled to `target/debug/examples/test` and appears to be built without the test attribute as running the executable directly doesn't output `cfg: test`

examples/test.rs:

```
fn main() {
    if cfg!(test) {
        println!("cfg: test");
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
    }
}
```

**Comment by alexcrichton** [comment]
Yeah currently we don't compile examples with `--test` I believe (they're intended to be examples of using a library), but this should be easy enough to change with a flag I think.

**Comment by WaDelma** [comment]
I would really really love if cargo ran any tests in examples by default. Currently example code can easily get desync by changes in the library (even making CI to run examples requires special additions).

This makes examples currently feel like second class citizens: I added example to library that I am contributing and after changes made by other contributors I realised that the example wasn't working and I had to fix it myself and created dummy test file to tests folder that just includes the example (which was later removed and the CI just runs the example now).

Just now I made custom derive for a library to generate quickcheck tests and I had copied an example to tests folder to see if it works with it. I then pushed that copied file to my [PR](https://github.com/sebcrozet/alga/pull/16) and then realised after few days that I had the duplicate file there. Now I am faced with conundrum: Do I just move the example to tests folder because I want those generated tests tested or do I create dummy include file to tests folder.
